src/main/java/bgu/spl/mics/application/passiveObjects/Diary.java

```diff
@@ -38,23 +63,33 @@ public void addReport(Report reportToAdd){
 	 * List of all the reports in the diary.
 	 * This method is called by the main method in order to generate the output.
 	 */
-	public void printToFile(String filename){
-		//TODO: Implement this
+	public void printToFile(String filename) throws IOException {
+		_lock.readLock().lock();
+		Gson gson = new Gson();
+		String reportsToPrint = gson.toJson(_diary);
+		PrintWriter writer = null;
+		Path file = Paths.get(filename);
+		Files.write(file, Collections.singleton(reportsToPrint), StandardCharsets.UTF_8);
+		_lock.readLock().unlock();
 	}
 
 	/**
 	 * Gets the total number of received missions (executed / aborted) be all the M-instances.
 	 * @return the total number of received missions (executed / aborted) be all the M-instances.
 	 */
 	public int getTotal(){
-		//TODO: Implement this
-		return 0;
+		return _total.get();
 	}
 
 	/**
 	 * Increments the total number of received missions by 1
 	 */
 	public void incrementTotal(){
-
+		int localTotal;
+		do
+		{
+			localTotal = _total.get();
+		}
+		while (!_total.compareAndSet(localTotal, localTotal+1));
```

Comment:
is this +1 safe? maybe save a  newTotal variable in the do loop

Reply:
It is run before calling compareAndSet, so if some thread comes in between them the cas will still fail. This won't help against a thread coming in between but it isn't different by any means than putting it inside the do-while